File: ws-server/src/main/java/net/devopssolutions/demo/ws/server/handler/WsHandler.java
package net.devopssolutions.demo.ws.server.handler;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.CloseStatus;
import org.springframework.web.socket.WebSocketHandler;
import org.springframework.web.socket.WebSocketMessage;
import org.springframework.web.socket.WebSocketSession;

@Slf4j
@Component
public class WsHandler implements WebSocketHandler {

    @Override
    public void afterConnectionEstablished(WebSocketSession session) throws Exception {
        log.info("afterConnectionEstablished session: {}", session);
    }

    @Override
    public void handleMessage(WebSocketSession session, WebSocketMessage<?> message) throws Exception {
        log.info("handleMessage session: {} message: {}", session, message);
    }

    @Override
    public void handleTransportError(WebSocketSession session, Throwable exception) throws Exception {
        log.info("handleTransportError session: " + session, exception);
    }

    @Override
    public void afterConnectionClosed(WebSocketSession session, CloseStatus closeStatus) throws Exception {
        log.info("afterConnectionClosed session: {} closeStatus: {}", session, closeStatus);
    }

    @Override
    public boolean supportsPartialMessages() {
        return false;
    }
}

File: ws-client/src/main/kotlin/net/devopssolutions/demo/ws/client/component/WsHandler.kt
package net.devopssolutions.demo.ws.client.component

import mu.KLogging
import org.springframework.stereotype.Component
import org.springframework.web.reactive.socket.WebSocketSession
import org.springframework.web.reactive.socket.client.WebSocketClient
import reactor.core.publisher.Mono
import reactor.core.publisher.SignalType
import reactor.util.Loggers
import java.net.URI
import java.time.Duration
import java.util.logging.Level
import javax.annotation.PreDestroy

val excludeOnNextAndRequest = SignalType.values().asSequence()
        .minus(SignalType.REQUEST)
        .minus(SignalType.ON_NEXT)
        .toList().toTypedArray()

@Component
class WsHandler(
        webSocketClient: WebSocketClient,
        private val wsConsumer: WsConsumer,
        private val wsProducer: WsProducer
) {
    companion object : KLogging()

    private val subscription = webSocketClient
            .execute(URI("ws://localhost:8080/ws"), ::registerHandlers)
            .retryWhen { it.flatMap { Mono.delay(Duration.ofSeconds(5)) } }
            .log()
            .subscribe()
    private val messageCountLogger = wsProducer.logMessagesCount(wsConsumer.messagesCountInSecond)
            .retry()
            .subscribe()

    private fun registerHandlers(session: WebSocketSession): Mono<Void> {
        logger.info("opened ws connection session: {}, endpointConfig: {}", session, session.handshakeInfo)

        val (emitter, sink) = wsProducer.buildSender(session)

        return session.send(emitter)
                .mergeWith(wsConsumer.createReceiver(session, sink).then())
                .log(Loggers.getLogger("session"), Level.INFO, true)
                .then()
    }

    @PreDestroy
    private fun stop() {
        subscription.dispose()
        messageCountLogger.dispose()
    }

}
